scipy/sparse/_dok.py

```diff
@@ -456,17 +461,25 @@ class dok_array(_dok_base, sparray):
 
 dok_array.__doc__ = _dok_base.__doc__
 
-class dok_matrix(spmatrix, _dok_base):
+class dok_matrix(spmatrix, _dok_base, dict):
     def set_shape(self, shape):
         new_matrix = self.reshape(shape, copy=False).asformat(self.format)
         self.__dict__ = new_matrix.__dict__
-        dict.clear(self)
-        dict.update(self, new_matrix)
 
     def get_shape(self):
         """Get shape of a sparse array."""
         return self._shape
 
     shape = property(fget=get_shape, fset=set_shape)
 
+    #######
+    # It is not clear that these should be defined -- not part of sparse array api
+    # but they are included for backward compatibility
+    def setdefault(self, key, default=None, /):
+        return self._dict.setdefault(key, default)
+
+    def __delitem__(self, key, /):
+        del self._dict[key]
```

Comment:
Couldn't it be pulled up in `_dok_base`?

Reply:
Yes -- `setdefault` and `__delitem__` could both be included in `_dok_base`. Doing that means they appear in both `dok_matrix` and `dok_array`.  The current placement means they are available in `dok_matrix` for backward compatibility but don't appear in `dok_array. I don't feel strongly about placing any of these in either class so I am interested in what you think makes the most sense. 

I was thinking that `__delitem__` and `setdefault` were strange methods for a sparse matrix. The `__delitem__` sets the value of that key to zero. It doesn't delete it. And `setdefault` is the same as `__setitem__` unless the current value is zero (not in the indices) -- in which case it sets the value to the provided default input.  I can't imagine a case for doing that in a sparse matrix. But I'm fine with putting either into both dok_array is others think that is better.

There may be others that I put in _dok_base that maybe should be moved to only be in sparse_matrix:
- `popitem` is a strange way to set the value to be zero and return the value that has been set to zero.
- `conjtransp` which seems to be a method that someone forgot to remove when it was removed from the others.
- I'm not sure what the right behavior is for `__contains__`. `np.array` uses contains to look for a **value** somewhere in the array. This `dok_matrix.__contains__` method returns `True` when the input is a **key** hs a nonzero value in the matrix. I left it in `dok_array` because it is a common dict-like pattern. But I'm not sure it should be provided. The other formats seem to raise exceptions with, e.g., `3 in A`. 

I can certainly put `__delitem__` into `dok_array` if that would be helpful. Thoughts on these others?

Reply:
I think I'd prefer to keep both `setdefault` and `__delitem__` in `_dok_base`, as it's not uncommon to change the sparsity structure of a DOK matrix by deleting particular indices via `del x[i, j]`. Using setdefault is more rare, but I can see its use cases too.

I'm not sure what you meant by "The `__delitem__` sets the value of that key to zero. It doesn't delete it." The current implementation of `dok_matrix` definitely deletes the key/value pair from the dictionary. In fact, the DOK format doesn't support storing explicit zeros at all.

I think `popitem` is fine to keep, and `conjtransp` should probably be deprecated (and then removed after 2 releases).

`__contains__` should continue to use dict semantics, as this is exactly what the DOK format was intended to be.